src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+
+  /* force these devices as deselected (auto state saves in db later) since 
+   * these need use to select (and cause the device probe to start connection to
+   * remote side
+   */
+  device->selected = 0;
+
+  s->callback_id = callback_id;
+  // tear this session down, incl free'ing it
+  rcp_session_shutdown(s, RCP_STATE_STOPPED);
+
+  return 1;
+}
+
+static int
+rcp_device_flush(struct output_device *device, int callback_id)
+{
+  DPRINTF(E_DBG, L_RCP, "device flush %ld '%s'\n", device->id, device->name);
+
+  // needed?
+  struct rcp_session *s = device->session;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static int
+rcp_device_probe(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device probe %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+
+static int
+rcp_device_volume_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+  char cmd[3];
+  int ret;
+
+  DPRINTF(E_DBG, L_RCP, "device volume set %ld '%s' state=%d vol=%d\n", device->id, device->name, s->state, device->volume);
+
+  if (!s || (s && s->state != RCP_STATE_STREAMING))
+    return 0;
+
+  s->callback_id = callback_id;
+
+
+  ret = snprintf(cmd, sizeof(cmd), "%d", device->volume);
+  if (ret < 0) {
+    return 0;
+  }
+
+  rcp_send(s, RCP_STATE_VOL_SET, cmd);
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static void
+rcp_device_cb_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device cb %ld '%s'  callback %d\n", device->id, device->name, callback_id);
+
+  s->callback_id = callback_id;
+}
+
+static void
+rcp_device_add(const char *device_name, const char *address)
+{
+  struct output_device *device;
+  uint64_t id;
+  int ret;
+
+  /* possible if this was mdns/avahi that there is a timeout/cached entry is no
+   * longer resolvable;  ie avahi caches a roku name, the roku is pwoered down,
+   * cache entry still exists but now no address
+   */ 
+  if (!address)
+    {
+      DPRINTF(E_DBG, L_RCP, "Attempt to add '%s' without address (possible mdns timeout)\n", device_name ? device_name : "<no name>");
+      return;
+    }
+
+  id = djb_hash(address, strlen(address));
+  if (!id)
+    {
+      DPRINTF(E_LOG, L_RCP, "Could not hash RCP adddress name for id (%s)\n", address);
+      return;
+    }
+
+  CHECK_NULL(L_RCP, device = calloc(1, sizeof(struct output_device)));
+
+  // 3.0.44 being the last public firmware of the Roku Soundbridge devices...
+  device->id = id + 3044;
+  device->name = strdup(device_name);
+  device->type = OUTPUT_TYPE_RCP;
+  device->type_name = outputs_name(device->type);
+  device->has_video = 0;
```

Comment:
Not necessary

Reply:
copy / paste from outputs/dummy.c, dropped